src/app/Library/CrudPanel/Traits/ColumnsProtectedMethods.php

```diff
@@ -80,14 +80,55 @@ protected function makeSureColumnHasLabel($column)
      */
     protected function makeSureColumnHasType($column)
     {
-        $could_be_relation = isset($column['entity']) && $column['entity'] !== false;
+        // Do not alter type if it has been set by developer
+        if (isset($column['type'])) {
+            return $column;
+        }
+
+        // Set text as default column type
+        $column['type'] = 'text';
+
+        $could_be_relation = Arr::get($column, 'entity', false) !== false;
 
-        if (! isset($column['type']) && $could_be_relation) {
+        if ($could_be_relation) {
             $column['type'] = 'relationship';
         }
 
-        if (! isset($column['type'])) {
-            $column['type'] = 'text';
+        if (in_array($column['name'], $this->model->getDates())) {
+            $column['type'] = 'datetime';
+        }
+
+        if ($this->model->hasCast($column['name'])) {
+            $attributeType = $this->model->getCasts()[$column['name']];
+
+            switch ($attributeType) {
+                case 'array':
+                case 'encrypted:array':
+                case 'collection':
+                case 'encrypted:collection':
+                case 'json':
```

Comment:
`json` should set `json` type column I guess ?

Reply:
Yeah, I think that would be for the better. Same for `object` then. We can't assume the a `json` casted attribute contains an array... it could just as well be an `object` or something. 

So I'd default to `json` column type for both `json` and `object`... **What do you think @jnoordsij ? Are we missing something here?**

Other than that... I love this, I see absolutely no other problems with it, I'd love to merge it into 4.2 right away. Thanks a lot! 🙏

Reply:
That seems to be a better choice indeed! Fixed it.